sjsonnet/src/sjsonnet/Val.scala

```diff
@@ -354,6 +455,124 @@ object Val{
       }
     }
   }
+
+  abstract class Builtin(paramNames: String*)
+    extends Func(null, null, Params(paramNames.toArray, new Array[Expr](paramNames.length), paramNames.indices.toArray)) {
+
+    final def evalRhs(scope: ValScope, ev: EvalScope, fs: FileScope, pos: Position): Val = {
+      val args = new Array[Val](params.names.length)
+      var i = 0
+      while(i < args.length) {
+        args(i) = scope.bindings(i).force
+        i += 1
+      }
+      evalRhs(args, ev, pos)
+    }
+
+    def evalRhs(args: Array[Val], ev: EvalScope, pos: Position): Val
+
+    override def apply(argNames: Array[String], argVals: Array[Lazy],
+              outerPos: Position)
+             (implicit ev: EvalScope): Val = {
+
+      val argsSize = argVals.length
+      val simple = argNames == null && params.indices.length == argsSize
+      val passedArgsBindingsI = if(argNames != null) {
+        val arrI: Array[Int] = new Array(argsSize)
+        var i = 0
+        try {
+          while (i < argsSize) {
+            val aname = argNames(i)
+            arrI(i) = if(aname != null) params.argIndices.getOrElse(
+              aname,
+              Error.fail(s"Function has no parameter $aname", outerPos)
+            ) else params.indices(i)
+            i += 1
+          }
+        } catch { case e: IndexOutOfBoundsException =>
+          Error.fail("Too many args, function has " + params.names.length + " parameter(s)", outerPos)
+        }
+        arrI
+      } else if(params.indices.length < argsSize) {
```

Comment:
Would moving this case to the top of the if-else chain allow us to avoid the `try`-`catch` above? That would also allow us to avoid duplicating the `Error.fail` call.